File: sdks/typescript/src/context.ts
import type { DeepPartial, WorkerRequest } from "./generated/valka/v1/worker.js";
import { LogLevel } from "./types.js";

type SendFn = (msg: DeepPartial<WorkerRequest>) => void;

export interface SignalData {
  signalId: string;
  name: string;
  payload: string;
}

interface SignalWaiter {
  name: string | null; // null = any signal
  resolve: (data: SignalData) => void;
}

export class TaskContext {
  readonly taskId: string;
  readonly taskRunId: string;
  readonly queueName: string;
  readonly taskName: string;
  readonly attemptNumber: number;

  private readonly rawInput: string;
  private readonly rawMetadata: string;
  private readonly sendFn: SendFn;
  private readonly signalBuffer: SignalData[] = [];
  private readonly signalWaiters: SignalWaiter[] = [];

  /** @internal */
  constructor(
    taskId: string,
    taskRunId: string,
    queueName: string,
    taskName: string,
    attemptNumber: number,
    input: string,
    metadata: string,
    sendFn: SendFn,
  ) {
    this.taskId = taskId;
    this.taskRunId = taskRunId;
    this.queueName = queueName;
    this.taskName = taskName;
    this.attemptNumber = attemptNumber;
    this.rawInput = input;
    this.rawMetadata = metadata;
    this.sendFn = sendFn;
  }

  /** Parse the task input JSON. */
  input<T = unknown>(): T {
    return JSON.parse(this.rawInput || "null") as T;
  }

  /** Parse the task metadata JSON. */
  metadata<T = Record<string, unknown>>(): T {
    return JSON.parse(this.rawMetadata || "{}") as T;
  }

  /** Wait for a signal with a specific name. Non-matching signals are buffered. */
  waitForSignal(name: string): Promise<SignalData> {
    // Check buffer first
    const idx = this.signalBuffer.findIndex((s) => s.name === name);
    if (idx >= 0) {
      const signal = this.signalBuffer.splice(idx, 1)[0];
      this.sendSignalAck(signal.signalId);
      return Promise.resolve(signal);
    }

    // Register waiter
    return new Promise<SignalData>((resolve) => {
      this.signalWaiters.push({ name, resolve });
    });
  }

  /** Wait for the next signal (any name). Checks buffer first. */
  receiveSignal(): Promise<SignalData> {
    if (this.signalBuffer.length > 0) {
      const signal = this.signalBuffer.shift()!;
      this.sendSignalAck(signal.signalId);
      return Promise.resolve(signal);
    }

    return new Promise<SignalData>((resolve) => {
      this.signalWaiters.push({ name: null, resolve });
    });
  }

  /** Parse a signal's JSON payload. */
  static parseSignalPayload<T = unknown>(signal: SignalData): T {
    return JSON.parse(signal.payload || "null") as T;
  }

  /** @internal — called by the worker to deliver signals to this context. */
  _deliverSignal(signal: {
    signalId: string;
    signalName: string;
    payload: string;
  }): void {
    const data: SignalData = {
      signalId: signal.signalId,
      name: signal.signalName,
      payload: signal.payload,
    };

    // Check if any waiter matches
    for (let i = 0; i < this.signalWaiters.length; i++) {
      const waiter = this.signalWaiters[i];
      if (waiter.name === null || waiter.name === data.name) {
        this.signalWaiters.splice(i, 1);
        this.sendSignalAck(data.signalId);
        waiter.resolve(data);
        return;
      }
    }

    // No waiter matched, buffer it
    this.signalBuffer.push(data);
  }

  private sendSignalAck(signalId: string): void {
    this.sendFn({ signalAck: { signalId } });
  }

  /** Log at INFO level. */
  log(message: string): void {
    this.logAtLevel(LogLevel.INFO, message);
  }

  /** Log at DEBUG level. */
  debug(message: string): void {
    this.logAtLevel(LogLevel.DEBUG, message);
  }

  /** Log at WARN level. */
  warn(message: string): void {
    this.logAtLevel(LogLevel.WARN, message);
  }

  /** Log at ERROR level. */
  error(message: string): void {
    this.logAtLevel(LogLevel.ERROR, message);
  }

  private logAtLevel(level: LogLevel, message: string): void {
    this.sendFn({
      logBatch: {
        entries: [
          {
            taskRunId: this.taskRunId,
            timestampMs: Date.now(),
            level: level as number,
            message,
            metadata: "",
          },
        ],
      },
    });
  }
}

File: sdks/go/context.go
package valka

import (
	"context"
	"encoding/json"
	"time"

	pb "github.com/valka-queue/valka/sdks/go/proto/valkav1"
)

// SignalData holds the data from a received signal.
type SignalData struct {
	SignalID string
	Name     string
	Payload  string
}

// ParsePayload parses the signal's JSON payload into the provided destination.
func (s *SignalData) ParsePayload(dest interface{}) error {
	if s.Payload == "" {
		return nil
	}
	return json.Unmarshal([]byte(s.Payload), dest)
}

// TaskContext is provided to task handlers with task metadata and logging.
// It embeds context.Context for native cancellation support.
type TaskContext struct {
	context.Context

	TaskID        string
	TaskRunID     string
	QueueName     string
	TaskName      string
	AttemptNumber int32
	RawInput      string
	RawMetadata   string

	sendFn   func(*pb.WorkerRequest)
	cancel   context.CancelFunc
	signalCh chan *pb.TaskSignal
	sigBuf   []*pb.TaskSignal
}

// Input parses the task input JSON into the provided destination.
// Returns nil if input is empty.
func (c *TaskContext) Input(dest interface{}) error {
	if c.RawInput == "" {
		return nil
	}
	return json.Unmarshal([]byte(c.RawInput), dest)
}

// Metadata parses the task metadata JSON into the provided destination.
// Returns nil if metadata is empty.
func (c *TaskContext) Metadata(dest interface{}) error {
	if c.RawMetadata == "" {
		return nil
	}
	return json.Unmarshal([]byte(c.RawMetadata), dest)
}

// Log sends an INFO-level log entry.
func (c *TaskContext) Log(message string) {
	c.logAtLevel(pb.LogLevel_LOG_LEVEL_INFO, message)
}

// Debug sends a DEBUG-level log entry.
func (c *TaskContext) Debug(message string) {
	c.logAtLevel(pb.LogLevel_LOG_LEVEL_DEBUG, message)
}

// Warn sends a WARN-level log entry.
func (c *TaskContext) Warn(message string) {
	c.logAtLevel(pb.LogLevel_LOG_LEVEL_WARN, message)
}

// Error sends an ERROR-level log entry.
func (c *TaskContext) Error(message string) {
	c.logAtLevel(pb.LogLevel_LOG_LEVEL_ERROR, message)
}

// WaitForSignal waits for a signal with the given name. Non-matching signals are buffered.
// Returns nil if the context is cancelled before a matching signal arrives.
func (c *TaskContext) WaitForSignal(name string) (*SignalData, error) {
	// Check buffer first
	for i, sig := range c.sigBuf {
		if sig.SignalName == name {
			c.sigBuf = append(c.sigBuf[:i], c.sigBuf[i+1:]...)
			c.sendSignalAck(sig.SignalId)
			return &SignalData{SignalID: sig.SignalId, Name: sig.SignalName, Payload: sig.Payload}, nil
		}
	}

	// Wait for matching signal from channel
	for {
		select {
		case <-c.Done():
			return nil, c.Err()
		case sig, ok := <-c.signalCh:
			if !ok {
				return nil, context.Canceled
			}
			if sig.SignalName == name {
				c.sendSignalAck(sig.SignalId)
				return &SignalData{SignalID: sig.SignalId, Name: sig.SignalName, Payload: sig.Payload}, nil
			}
			c.sigBuf = append(c.sigBuf, sig)
		}
	}
}

// ReceiveSignal waits for the next signal (any name). Checks buffer first.
func (c *TaskContext) ReceiveSignal() (*SignalData, error) {
	// Check buffer first
	if len(c.sigBuf) > 0 {
		sig := c.sigBuf[0]
		c.sigBuf = c.sigBuf[1:]
		c.sendSignalAck(sig.SignalId)
		return &SignalData{SignalID: sig.SignalId, Name: sig.SignalName, Payload: sig.Payload}, nil
	}

	select {
	case <-c.Done():
		return nil, c.Err()
	case sig, ok := <-c.signalCh:
		if !ok {
			return nil, context.Canceled
		}
		c.sendSignalAck(sig.SignalId)
		return &SignalData{SignalID: sig.SignalId, Name: sig.SignalName, Payload: sig.Payload}, nil
	}
}

func (c *TaskContext) sendSignalAck(signalID string) {
	c.sendFn(&pb.WorkerRequest{
		Request: &pb.WorkerRequest_SignalAck{
			SignalAck: &pb.SignalAck{
				SignalId: signalID,
			},
		},
	})
}

func (c *TaskContext) logAtLevel(level pb.LogLevel, message string) {
	entry := &pb.LogEntry{
		TaskRunId:   c.TaskRunID,
		TimestampMs: time.Now().UnixMilli(),
		Level:       level,
		Message:     message,
	}
	c.sendFn(&pb.WorkerRequest{
		Request: &pb.WorkerRequest_LogBatch{
			LogBatch: &pb.LogBatch{
				Entries: []*pb.LogEntry{entry},
			},
		},
	})
}
